scripts/enforce_match_arg_in_assert_produces_warning.py

```diff
@@ -0,0 +1,81 @@
+"""
+
+Enforce that all usages of tm.assert_produces_warning use
+the "match" argument. This will help ensure that users always see
+the correct warning message.
+
+tm.assert_produces_warning(None) is excluded as no warning is
+produced.
+
+"""
+
+
+import argparse
+import ast
+from collections.abc import Sequence
+import sys
+
+ERROR_MESSAGE = (
+    "{path}:{lineno}:{col_offset}: "
+    '"match" argument missing in tm.assert_produces_warning'
+    "\n"
+)
+
+
+class MatchArgForWarningsChecker(ast.NodeVisitor):
+    def __init__(self) -> None:
+        self.error_set = []
+
+    def visit_Call(self, node) -> None:
+        if ( isinstance(node.func, ast.Attribute)  and
+            node.func.attr == "assert_produces_warning"):
+            # only check for attribute function of class/module tm
+            if ( isinstance(node.func.value, ast.Name) and
+                node.func.value.id == "tm" ):
+                # ignore tm.assert_produces_warning(None)/tm.assert_produces_warning()
+                if ( len(node.args) == 0 or
+                    (isinstance(node.args[0], ast.Constant) and
+                    node.args[0].value is None) ):
```

Comment:
I believe `assert_produces_warning(False)` needs to be filtered as well, it has the same behavior as passing `None`